source/preprocess.d

```diff
@@ -0,0 +1,140 @@
+module preprocess;
+import language;
+import std.stdio;
+
+int depth;
+void replaceFor(ref Statement[] list, int t)
+{
+    auto f = cast(ForBlock) list[t];
+    if (f.style == ForStyle.until)
+    {
+        Statement[] b = list[0 .. t], e = list[t + 1 .. list.length];
+        auto mod = new BinaryOpExpression(BinaryOpExpression.Type.modulo, new BinaryOpExpression(
+                BinaryOpExpression.Type.subtract, f.finish, f.start), new ConstExpression(
+                depth));
+        auto it = new VarExpression(f.name, null);
+        auto cond1 = new BinaryOpExpression(BinaryOpExpression.Type.less, it, new BinaryOpExpression(
+                BinaryOpExpression.Type.add, f.start, mod));
+        auto cond2 = new BinaryOpExpression(BinaryOpExpression.Type.less, it, f.finish);
+        auto incrIt = new AssignStatement(f.lineId, AssignStatement.Type.assignAdd, it, new ConstExpression(
+                1));
+        auto while1 = new WhileBlock(f.lineId, cond1);
+        while1.statementList = f.statementList;
+        while1.statementList ~= incrIt;
+        auto while2 = new WhileBlock(f.lineId, cond2);
+        for (int i = 0; i < depth; i++)
+        {
+            while2.statementList ~= f.statementList;
+            while2.statementList ~= incrIt;
+        }
+        b ~= new AssignStatement(f.lineId, AssignStatement.Type.assign, it, f.start);
+        b ~= while1;
+        b ~= while2;
+        b ~= e;
+        list = b;
+    }
+    if (f.style == ForStyle.rangeto)
```

Comment:
I don't believe these are actually three __entirely different cases__. Please extract the repeating body into a function, and call it three times with appropriate parameters.